RecoJets/JetProducers/src/PileupJetIdAlgo.cc

```diff
@@ -139,52 +140,42 @@ void setPtEtaPhi(const reco::Candidate & p, float & pt, float & eta, float &phi
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
             }
+            for(std::vector<std::string>::iterator it=tmvaSpectators_.begin(); it!=tmvaSpectators_.end(); ++it) {
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
+            }
+            std::unique_ptr<TMVA::IMethod> temp( reco::details::loadTMVAWeights(&tmpTMVAReader,  tmvaMethod_.c_str(), tmvaEtaWeights_.at(v).c_str() ) );
+            etaReader_.push_back(std::unique_ptr<const GBRForest> ( new GBRForest( dynamic_cast<TMVA::MethodBDT*>( tmpTMVAReader.FindMVA(tmvaMethod_.c_str()) ) ) ) );
           }
 	} else {
-	  for(std::vector<std::string>::iterator it=tmvaVariables_.begin(); it!=tmvaVariables_.end(); ++it) {
-		if(  tmvaNames_[*it].empty() ) { 
-			tmvaNames_[*it] = *it;
-		}
-		reader_->AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
-	  }
-	}
-	for(std::vector<std::string>::iterator it=tmvaSpectators_.begin(); it!=tmvaSpectators_.end(); ++it) {
-		if(  tmvaNames_[*it].empty() ) { 
-			tmvaNames_[*it] = *it;
-		}
-		if(etaBinnedWeights_){
-                  for(int v=0; v<nEtaBins_;v++){
-                        etaReader_.at(v)->AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
-                  }
-		} else {
-			reader_->AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
-		}
-	}
-	if(etaBinnedWeights_){
-          for(int v=0; v<nEtaBins_;v++){
-                reco::details::loadTMVAWeights(etaReader_.at(v).get(),  tmvaMethod_.c_str(), tmvaEtaWeights_.at(v).c_str() );
-          }
-	} else {
-		reco::details::loadTMVAWeights(reader_.get(),  tmvaMethod_.c_str(), tmvaWeights_.c_str() ); 
+            TMVA::Reader tmpTMVAReader( "!Color:Silent:!Error" );
+            for(std::vector<std::string>::iterator it=tmvaVariables_.begin(); it!=tmvaVariables_.end(); ++it) {
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
+            }
+            for(std::vector<std::string>::iterator it=tmvaSpectators_.begin(); it!=tmvaSpectators_.end(); ++it) {
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
+            }
+            std::unique_ptr<TMVA::IMethod> temp( reco::details::loadTMVAWeights(&tmpTMVAReader,  tmvaMethod_.c_str(), tmvaWeights_.c_str() ) );
+            reader_ = std::unique_ptr<const GBRForest> ( new GBRForest( dynamic_cast<TMVA::MethodBDT*>( tmpTMVAReader.FindMVA(tmvaMethod_.c_str()) ) ) );
```

Comment:
Lines 164-178 mostly duplicate lines 147-161. Please eliminate this duplication by adding a method that can be called for each part of the `if-else`. I think the following should work:
```
std::unique_ptr<const GBRForest> PileupJetIdAlgo::getMVA(const std::vector<std::string> &varList, const std::string &tmvaWeights)
{
    TMVA::Reader tmpTMVAReader( "!Color:Silent:!Error" );
    for(std::vector<std::string>::iterator it=varList.begin(); it!=varList.end(); ++it) {
        if(  tmvaNames_[*it].empty() ) {
                tmvaNames_[*it] = *it;
        }
        tmpTMVAReader.AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
    }
    for(std::vector<std::string>::iterator it=tmvaSpectators_.begin(); it!=tmvaSpectators_.end(); ++it) {
        if(  tmvaNames_[*it].empty() ) {
                tmvaNames_[*it] = *it;
        }
        tmpTMVAReader.AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
    }
    std::unique_ptr<TMVA::IMethod> temp( reco::details::loadTMVAWeights(&tmpTMVAReader,  tmvaMethod_.c_str(), tmvaWeights.c_str() ) );
    return( std::make_unique<const GBRForest> ( dynamic_cast<TMVA::MethodBDT*>( tmpTMVAReader.FindMVA(tmvaMethod_.c_str()) ) ) );
}


void PileupJetIdAlgo::bookReader()
{
        if(etaBinnedWeights_){
          for(int v=0; v<nEtaBins_;v++) {
            etaReader_.push_back(getMVA(tmvaEtaVariables_.at(v), tmvaEtaWeights_.at(v)));
          }
        } else {
            reader_ = getMVA(tmvaVariables_, tmvaWeights_);
        }
}
```